### PR TITLE
Feature/secret override

### PR DESCRIPTION
**Pull request by cberg-aot**
# Description

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: charts/coms/templates/secret.yaml [hunk 1]
```diff
@@ -13,26 +21,62 @@
...
     "helm.sh/resource-policy": keep
+  {{- end }}
   name: {{ $bSecretName }}
   labels: {{ include "coms.labels" . | nindent 4 }}
 type: kubernetes.io/basic-auth
 data:
-  password: {{ $bPassword }}
-  username: {{ $bUsername }}
+  password: {{ .Values.basicAuthSecretOverride.password  | default $bPassword | quote }}
+  username: {{ .Values.basicAuthSecretOverride.username  | default $bUsername | quote }}
```
Comment (jujaga) [inline charts/coms/templates/secret.yaml:33]: Style: extra whitespace before the |

File: charts/coms/templates/secret.yaml [hunk 2]
```diff
@@ -13,26 +21,62 @@
...
     "helm.sh/resource-policy": keep
+  {{- end }}
   name: {{ $pSecretName }}
   labels: {{ include "coms.labels" . | nindent 4 }}
 type: kubernetes.io/basic-auth
 data:
-  password: {{ $pPassword }}
-  username: {{ $pUsername }}
+  password: {{ .Values.dbSecretOverride.password  | default $pPassword | quote }}
+  username: {{ .Values.dbSecretOverride.username  | default $pUsername | quote }}
```
Comment (jujaga) [inline charts/coms/templates/secret.yaml:49]: Style: extra whitespace before the |

File: charts/coms/templates/secret.yaml [hunk 3]
```diff
@@ -13,26 +21,62 @@
...
+  {{- if not .Values.config.releaseScoped }}
+  annotations:
+    "helm.sh/resource-policy": keep
+  {{- end }}
+  name: {{ $oSecretName }}
+  labels: {{ include "coms.labels" . | nindent 4 }}
+type: kubernetes.io/basic-auth
+data:
+  password: {{ .Values.objectStorageSecretOverride.password  | default $oPassword | quote }}
+  username: {{ .Values.objectStorageSecretOverride.username  | default $oUsername | quote }}
```
Comment (jujaga) [inline charts/coms/templates/secret.yaml:65]: Style: extra whitespace before the |

**Review by jujaga — DISMISSED** [review]
On visual inspection this looks correct. Other than a minor whitespace style issue, this looks good otherwise. Please tweak that, and make sure to rebase branch to the latest upstream changes (PR-148 just got merged in this morning) and it should be good to go. 👍 
Pre-emptively signing off on this with the acknowledgement that the minor style changes will be done.

**Review by jujaga — CHANGES_REQUESTED** [review]
Just thought through your PR a bit more. I think the original reason why we didn't include the `keycloak` and `objectstorage` secret definitions was that it does not really ever make sense to auto-generate them as they're externally defined secrets. Would you be able to modify your PR so that in addition to checking if the secret does not already exist, to only execute and apply those two secrets IF and ONLY IF you have the respective override values defined in your values? (Ref `charts/coms/templates/secret.yaml` Lines 51 and 67 for where I think we'll need more conditional checks)

**Comment by cberg-aot** [comment]
I did a manual smell check of the conditional templating logic.  

I played with logic:
 - (or username password) 
 - (and username password) 
Autogenerating a default password/username could still be supported using or gate.

File: charts/coms/templates/secret.yaml [hunk 4]
```diff
@@ -1,10 +1,18 @@
 {{- $bPassword := (randAlphaNum 32) | b64enc }}
 {{- $bUsername := (randAlphaNum 32) | b64enc }}
+{{- $kPassword := (randAlphaNum 32) | b64enc }}
+{{- $kUsername := (randAlphaNum 32) | b64enc }}
+{{- $oPassword := (randAlphaNum 32) | b64enc }}
+{{- $oUsername := (randAlphaNum 32) | b64enc }}
```
Comment (jujaga) [inline charts/coms/templates/secret.yaml:6]: We can probably drop these as I do not believe there are any circumstances where we would ever want to use this.

File: charts/coms/templates/secret.yaml [hunk 5]
```diff
@@ -13,26 +21,62 @@
...
+  {{- if not .Values.config.releaseScoped }}
+  annotations:
+    "helm.sh/resource-policy": keep
+  {{- end }}
+  name: {{ $oSecretName }}
+  labels: {{ include "coms.labels" . | nindent 4 }}
+type: kubernetes.io/basic-auth
+data:
+  password: {{ .Values.objectStorageSecretOverride.password | default $oPassword | quote }}
+  username: {{ .Values.objectStorageSecretOverride.username | default $oUsername | quote }}
```
Comment (jujaga) [inline charts/coms/templates/secret.yaml:65]: We can drop the default generated fallback clauses ans just pipe quote the values here.

File: charts/coms/templates/secret.yaml [hunk 6]
```diff
@@ -13,26 +21,62 @@
...
+  {{- if not .Values.config.releaseScoped }}
+  annotations:
+    "helm.sh/resource-policy": keep
+  {{- end }}
+  name: {{ $kSecretName }}
+  labels: {{ include "coms.labels" . | nindent 4 }}
+type: kubernetes.io/basic-auth
+data:
+  password: {{ .Values.keycloakSecretOverride.password | default $kPassword | quote }}
+  username: {{ .Values.keycloakSecretOverride.username | default $kUsername | quote }}
```
Comment (jujaga) [inline charts/coms/templates/secret.yaml:81]: We can drop the default generated fallback clauses ans just pipe quote the values here.

**Review by jujaga — CHANGES_REQUESTED** [review]
Thanks for the general logic wrapper changes. I believe the and path is correct as we need to enforce the following:
1) If you already have a secret, do not touch it
2) If you don't have a secret, and you do have values specified, use them
The auto-generating behavior, while nice to have, is completely superfluous now with this logic flow and should be trimmed out.

File: charts/coms/templates/secret.yaml (hunk 6 above)
Comment (cberg-aot) [reply charts/coms/templates/secret.yaml]: Refactored and gave it a sanity install for good measure.

**Review by jujaga — APPROVED** [review]

**Review by kamorel — APPROVED** [review]

**Pull request merged**